### ISSUE TITLE
Replace constants that use `#define` with `constexpr`

### ISSUE THREAD

**Issue by Gymnasiast** [OpenRCT2/OpenRCT2#21421]
Since switching to C++, we declare constants using `constexpr`. However, there are plenty of older constants from our C days. These need to be refactored.

Actual macros (which are also declared using `#define` fall outside the scope of this issue).

**Comment by janclod** [comment]
First time contributor here, can I be of any help here? Or is it better if one person tackle this on its own?

**Comment by tupaschoal** [comment]
> First time contributor here, can I be of any help here? Or is it better if one person tackle this on its own?

There is definitely room for more than one person here. Find one define that has moderate usage and comment here that you'll be tackling it. Once you're done, open a PR. That will help you get the hang of the repo, on next PRs you might try more commits to handle bigger and more complex defines, but the key is being organized and keeping PRs size relatively under control for faster reviews.

**Comment by janclod** [comment]
I started experimenting :) this is how it looks like:
https://github.com/OpenRCT2/OpenRCT2/compare/develop...janclod:OpenRCT2:replace-define-with-constexpr

In this example I took RIDE_ADJACENCY_CHECK_DISTANCE inside Ride.h

Any feedback welcome...

**Comment by Gymnasiast** [comment]
The first commit looks good, haven’t taken a look at the other ones. I do recommend you create a PR now before your branch gets too big. This makes it easier to review and easier to make changes, should they be necessary.

**Comment by janclod** [comment]
How should I move on from here? Keep making separate PRs for every 5-10 replacement (of #defines) I do?

**Comment by tupaschoal** [comment]
> How should I move on from here? Keep making separate PRs for every 5-10 replacement (of #defines) I do?

Yes, don't make it so much as a fixed number of commits, think about their overall size.

**Comment by janclod** [comment]
Do we want to replace also this kind of `#define` with constexpr?

![image](https://github.com/OpenRCT2/OpenRCT2/assets/6673668/f336ccae-5e83-44c0-ac52-ada166cce31f)

**Comment by tupaschoal** [comment]
> Do we want to replace also this kind of `#define` with `constexpr`?
> 
> ![image](https://private-user-images.githubusercontent.com/6673668/313476226-f336ccae-5e83-44c0-ac52-ada166cce31f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA3MDkyMDYsIm5iZiI6MTcxMDcwODkwNiwicGF0aCI6Ii82NjczNjY4LzMxMzQ3NjIyNi1mMzM2Y2NhZS01ZTgzLTQ0YzAtYWM1Mi1hZGExNjZjY2UzMWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTdUMjA1NTA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjU4OTRkZDc5NmI5NzhlZTc0NzY4ZjMxZjAxMDBiNjUzMGQzYWNiNTNlZTViMWU2YWJmMDMxM2JlZTRjYzcyOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.w6uFDzhE-M28SgXAZ-2TxAQViFoWXuUYt0lgjWu9WTU)

yes

**Comment by tupaschoal** [comment]
Actually this one would be even better as a user-defined operator, like we did for money.

**Comment by janclod** [comment]
Do we want to replace this kind of define? Is it even possible?

![image](https://github.com/OpenRCT2/OpenRCT2/assets/6673668/aa41da5b-af31-4b01-b228-f699f4834d3d)

**Comment by janclod** [comment]
I am trying to replace this `define`:
![image](https://github.com/OpenRCT2/OpenRCT2/assets/6673668/14ed5af0-704e-4f8b-ae27-1242f76828a8)
inside `Theme.cpp`

AFAIK I would need to return a struct containing `uint8_t` and `WindowTheme` but there is no such struct. Also it is not allowed to return two types from one function.

Options:
1. create a struct so that we can maintain the `#define` unchanged
2. hardcode the `uint8_t` and replace part of the define, the one which returns `WindowTheme`
3. create two separate constexpr function

I would choose option 1, feedback welcome...

**Comment by Gymnasiast** [comment]
> Do we want to replace this kind of define? Is it even possible?
> 
> ![image](https://private-user-images.githubusercontent.com/6673668/317207441-aa41da5b-af31-4b01-b228-f699f4834d3d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1NTQxODgsIm5iZiI6MTcxMTU1Mzg4OCwicGF0aCI6Ii82NjczNjY4LzMxNzIwNzQ0MS1hYTQxZGE1Yi1hZjMxLTRiMDEtYjIyOC1mNjk5ZjQ4MzRkM2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjdUMTUzODA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2VjMDQ5ZjlhYTkxOWQ5Yjc0MDVjYjE4YzE4OWJjY2M3MTVhZWMxZWViODM2YTVkMTQwYzIyMTRkZWI2MDVhYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.TYCdBoA7WjVedfZjfDY-BDk0eESjpqqPsfqNrKPij4M)

Just leave it as-is, the scope of this is _constants_, rather than actual macros.

**Comment by Gymnasiast** [comment]
> I am trying to replace this `define`: ![image](https://private-user-images.githubusercontent.com/6673668/317327362-14ed5af0-704e-4f8b-ae27-1242f76828a8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1NTQxODgsIm5iZiI6MTcxMTU1Mzg4OCwicGF0aCI6Ii82NjczNjY4LzMxNzMyNzM2Mi0xNGVkNWFmMC03MDRlLTRmOGItYWUyNy0xMjQyZjc2ODI4YTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjdUMTUzODA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjJiODUxMjVlMzVlZGZjMTcyMzg1ZjRlY2NmM2FhMjg5NmU0ZjRmMWE5OWIzNjZiNmJlMjQ1ZjNiZGQ0ZjgzZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.CCM0mls12P0qTFGKckSe-4AcYq6m5zqmPtvt0BlsW1I) inside `Theme.cpp`
> 
> AFAIK I would need to return a struct containing `uint8_t` and `WindowTheme` but there is no such struct. Also it is not allowed to return two types from one function.
> 
> Options:
> 
>     1. create a struct so that we can maintain the `#define` unchanged
> 
>     2. hardcode the `uint8_t` and replace part of the define, the one which returns `WindowTheme`
> 
>     3. create two separate constexpr function
> 
> 
> I would choose option 1, feedback welcome...

I might be misreading it, but there is no point introducing a struct _and_ maintain a define. I would pick between leaving everything as-is _or_ creating a struct and adding constructors for all four used definitions (removing COLOURS_5 and 6, we can add those if and when they are actually needed). If you go down that path, you can also remove `COLOURS_RCT1` and just unwrap that, it doesn’t do anything to really clean up things.

**Comment by janclod** [comment]
> > I am trying to replace this `define`: ![image](https://private-user-images.githubusercontent.com/6673668/317327362-14ed5af0-704e-4f8b-ae27-1242f76828a8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1NTQxODgsIm5iZiI6MTcxMTU1Mzg4OCwicGF0aCI6Ii82NjczNjY4LzMxNzMyNzM2Mi0xNGVkNWFmMC03MDRlLTRmOGItYWUyNy0xMjQyZjc2ODI4YTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjdUMTUzODA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjJiODUxMjVlMzVlZGZjMTcyMzg1ZjRlY2NmM2FhMjg5NmU0ZjRmMWE5OWIzNjZiNmJlMjQ1ZjNiZGQ0ZjgzZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.CCM0mls12P0qTFGKckSe-4AcYq6m5zqmPtvt0BlsW1I) inside `Theme.cpp`
> > AFAIK I would need to return a struct containing `uint8_t` and `WindowTheme` but there is no such struct. Also it is not allowed to return two types from one function.
> > Options:
> > ```
> > 1. create a struct so that we can maintain the `#define` unchanged
> > 
> > 2. hardcode the `uint8_t` and replace part of the define, the one which returns `WindowTheme`
> > 
> > 3. create two separate constexpr function
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I would choose option 1, feedback welcome...
> 
> I might be misreading it, but there is no point introducing a struct _and_ maintain a define. I would pick between leaving everything as-is _or_ creating a struct and adding constructors for all four used definitions (removing COLOURS_5 and 6, we can add those if and when they are actually needed). If you go down that path, you can also remove `COLOURS_RCT1` and just unwrap that, it doesn’t do anything to really clean up things.

You are not misreading, just rookie proposal from my side :sweat_smile: 

I will see if I can add a struct with constructors...

**Comment by janclod** [comment]
> > > I am trying to replace this `define`: ![image](https://private-user-images.githubusercontent.com/6673668/317327362-14ed5af0-704e-4f8b-ae27-1242f76828a8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1NTQxODgsIm5iZiI6MTcxMTU1Mzg4OCwicGF0aCI6Ii82NjczNjY4LzMxNzMyNzM2Mi0xNGVkNWFmMC03MDRlLTRmOGItYWUyNy0xMjQyZjc2ODI4YTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjdUMTUzODA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjJiODUxMjVlMzVlZGZjMTcyMzg1ZjRlY2NmM2FhMjg5NmU0ZjRmMWE5OWIzNjZiNmJlMjQ1ZjNiZGQ0ZjgzZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.CCM0mls12P0qTFGKckSe-4AcYq6m5zqmPtvt0BlsW1I) inside `Theme.cpp`
> > > AFAIK I would need to return a struct containing `uint8_t` and `WindowTheme` but there is no such struct. Also it is not allowed to return two types from one function.
> > > Options:
> > > ```
> > > 1. create a struct so that we can maintain the `#define` unchanged
> > > 
> > > 2. hardcode the `uint8_t` and replace part of the define, the one which returns `WindowTheme`
> > > 
> > > 3. create two separate constexpr function
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I would choose option 1, feedback welcome...
> > 
> > 
> > I might be misreading it, but there is no point introducing a struct _and_ maintain a define. I would pick between leaving everything as-is _or_ creating a struct and adding constructors for all four used definitions (removing COLOURS_5 and 6, we can add those if and when they are actually needed). If you go down that path, you can also remove `COLOURS_RCT1` and just unwrap that, it doesn’t do anything to really clean up things.
> 
> You are not misreading, just rookie proposal from my side 😅
> 
> I will see if I can add a struct with constructors...

I had a look at this again and tried this possible implementation:
![image](https://github.com/OpenRCT2/OpenRCT2/assets/6673668/81f9d20b-c5f8-456c-801e-64280a63b7e6)

This also means that the definition of `WindowThemeDescriptors[]` will loose its `constexpr`...

What do you think? Does it make sense to refactor in this way? Do we gain much?

**Comment by rohan4412** [comment]
Another first timer here, is this still open to multiple people?

**Comment by tupaschoal** [comment]
Yes, just make sure to comment here and maybe align with @janclod so you don't do the same work

**Comment by janclod** [comment]
@rohan4412 there is still work to do, did you already start working on something?

**Comment by janclod** [comment]
> > > > I am trying to replace this `define`: ![image](https://private-user-images.githubusercontent.com/6673668/317327362-14ed5af0-704e-4f8b-ae27-1242f76828a8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1NTQxODgsIm5iZiI6MTcxMTU1Mzg4OCwicGF0aCI6Ii82NjczNjY4LzMxNzMyNzM2Mi0xNGVkNWFmMC03MDRlLTRmOGItYWUyNy0xMjQyZjc2ODI4YTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjdUMTUzODA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjJiODUxMjVlMzVlZGZjMTcyMzg1ZjRlY2NmM2FhMjg5NmU0ZjRmMWE5OWIzNjZiNmJlMjQ1ZjNiZGQ0ZjgzZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.CCM0mls12P0qTFGKckSe-4AcYq6m5zqmPtvt0BlsW1I) inside `Theme.cpp`
> > > > AFAIK I would need to return a struct containing `uint8_t` and `WindowTheme` but there is no such struct. Also it is not allowed to return two types from one function.
> > > > Options:
> > > > ```
> > > > 1. create a struct so that we can maintain the `#define` unchanged
> > > > 
> > > > 2. hardcode the `uint8_t` and replace part of the define, the one which returns `WindowTheme`
> > > > 
> > > > 3. create two separate constexpr function
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > I would choose option 1, feedback welcome...
> > > 
> > > 
> > > I might be misreading it, but there is no point introducing a struct _and_ maintain a define. I would pick between leaving everything as-is _or_ creating a struct and adding constructors for all four used definitions (removing COLOURS_5 and 6, we can add those if and when they are actually needed). If you go down that path, you can also remove `COLOURS_RCT1` and just unwrap that, it doesn’t do anything to really clean up things.
> > 
> > 
> > You are not misreading, just rookie proposal from my side 😅
> > I will see if I can add a struct with constructors...
> 
> I had a look at this again and tried this possible implementation: ![image](https://private-user-images.githubusercontent.com/6673668/317704382-81f9d20b-c5f8-456c-801e-64280a63b7e6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2OTA3NTgsIm5iZiI6MTcxMjY5MDQ1OCwicGF0aCI6Ii82NjczNjY4LzMxNzcwNDM4Mi04MWY5ZDIwYi1jNWY4LTQ1NmMtODAxZS02NDI4MGE2M2I3ZTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDlUMTkyMDU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWM3MjY4MjU2MGQyYmQzY2QwZWM2N2UzNTFkZjEzMGNkOTQ0YWJlZDAxMjVkMjg3YzQ4YzcwMDBmZmYyZmNjYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.n2zwAEbvBNUEllC76rr7cRAdw7BcO44KWKYf-cDZIbg)
> 
> This also means that the definition of `WindowThemeDescriptors[]` will loose its `constexpr`...
> 
> What do you think? Does it make sense to refactor in this way? Do we gain much?

Hey... Has anyone feedback on this? I personally like this approach and would like to implement all needed constructors for `NumColours`, but only after I got some positive feedback from one of you... Thanks in advance!

**Comment by rohan4412** [comment]
@janclod Yes I'd love to align, I haven't started working on anything just yet.

**Comment by janclod** [comment]
@rohan4412 are you in the Discord channel? Probably it is easiest there... What do you think?

I see there is also more activities related to refactoring `#define`, see #21818 and #21796

**Comment by rohan4412** [comment]
@janclod Yes I am, that sounds good. I am rohan4444 on Discord

**Comment by gwburk** [comment]
First time contributor, would it be OK for me to contribute here as well?

**Comment by janclod** [comment]
Welcome @gwburk :)

Yes there is still work to do, just keep in mind that there are already two people working on this (myself and @rohan4412).

In case you did not do it yet, please read the HowToContribute [guide](https://github.com/OpenRCT2/OpenRCT2/wiki/How-To-Contribute) and then you should be ready to go.

To avoid doing double work, I suggest you to coordinate your work with others on the Discord channel. Is that an option for you?

**Comment by gwburk** [comment]
Yes, I'm also gwburk on Discord.

**Comment by janclod** [comment]
> Yes, I'm also gwburk on Discord.

Great, I replied on Discord with overview of WIP.

**Comment by katerid** [comment]
Hi! I'm a first time contributor as well and was hoping that there would be room for be contribute here, would that be okay?

**Comment by janclod** [comment]
Welcome @katerid, yes there is still room :)

To avoid doing double work I suggest you to check Discord, there is a bit more information about possible starting point for new comers to this issue.

Also in case you did not find it yet, [here](https://github.com/OpenRCT2/OpenRCT2/wiki/How-To-Contribute) you find some general guidelines for first contributors.

**Comment by janclod** [comment]
I am moving here (from Discord) the overview of files worked on related to this issue.

TODO refactor:
FontFamilies.h
Color.h
Chat.h
Peep.h
Guest.h
common.h
SpriteCommands.cpp
CommandLine.hpp
audio.h
Version.h <- being worked on by katerid https://github.com/katerid/OpenRCT2

**Comment by jason-mills** [comment]
Another first time contributor here. Doesn't seem like anyone is actively working on this right now. Would it be alright for me to start work here?

**Comment by tupaschoal** [comment]
A lot was done, but there's still work to be done, when you find what you're doing, post here!

**Comment by jason-mills** [comment]
Looks like janclod's TODO list is still in need of work so I'll just start at the top there with FontFamilies and work my way down.

**Comment by tupaschoal** [comment]
Don't let the PR become too big before opening. It's better for us to have a few smaller PRs than a big one
